streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2002,58 +2002,41 @@ abstract class ZStream[-R, +E, +O](val process: ZManaged[R, Nothing, ZIO[R, Opti
-      }
+  )(l: O => O3, r: O2 => O3): ZStream[R1, E1, O3] =
+    ZStream {
+      import TerminationStrategy.{ Left => L, Right => R, Either => E }
 
-    combineChunks(that)((false, false, Option.empty[Loser])) {
-      case ((leftDone, rightDone, loser), left, right) =>
-        if (leftDone) {
-          right.map(c => (c.map(r), (leftDone, rightDone, None))).run
-        } else if (rightDone) {
-          left.map(c => (c.map(l), (leftDone, rightDone, None))).run
-        } else {
-          val result = loser match {
-            case None               => race(left.run, right.run)
-            case Some(Left(loser))  => race(loser.join, right.run)
-            case Some(Right(loser)) => race(left.run, loser.join)
-          }
-          result.flatMap {
-            case (exit, loser) =>
-              exit.foldM(
-                Cause.sequenceCauseOption(_) match {
-                  case Some(e) =>
-                    loser.merge.interrupt.as(Exit.halt(e.map(Some(_))))
-                  case None =>
-                    loser.fold(
-                      termination[E, O](leftDone, true, l),
-                      termination[E1, O2](true, rightDone, r)
-                    )
-                },
-                chunk => ZIO.succeedNow(Exit.succeed((chunk, (leftDone, rightDone, Some(loser)))))
-              )
-          }
-        }
+      for {
+        handoff <- ZStream.Handoff.make[Take[E1, O3]].toManaged_
+        done    <- Ref.makeManaged[Option[Boolean]](None)
+        chunksL <- self.process
+        chunksR <- that.process
+        handler = (pull: Pull[R1, E1, O3], terminate: Boolean) =>
+          done.get.flatMap {
+            case Some(true) =>
+              ZIO.succeedNow(false)
+            case _ =>
+              pull.run.flatMap { exit =>
+                done.modify {
+                  (_, exit.fold(c => Left(Cause.sequenceCauseOption(c)), Right(_))) match {
+                    case (state @ Some(true), _) =>
+                      ZIO.succeedNow(false) -> state
+                    case (state, Right(chunk)) =>
+                      handoff.offer(Exit.succeed(chunk)).as(true) -> state
+                    case (_, Left(Some(cause))) =>
+                      handoff.offer(Exit.halt(cause.map(Some(_)))).as(false) -> Some(true)
+                    case (option, Left(None)) if terminate || option.isDefined =>
+                      handoff.offer(Exit.fail(None)).as(false) -> Some(true)
+                    case (None, Left(None)) =>
+                      ZIO.succeedNow(false) -> Some(false)
+                  }
+                }.flatten
+              }
+          }.repeat(Schedule.doWhileEquals(true))
+        _ <- handler(chunksL.map(_.map(l)), List(L, E).contains(strategy)).fork.toManaged_
+        _ <- handler(chunksR.map(_.map(r)), List(R, E).contains(strategy)).fork.toManaged_
+      } yield handoff.take.flatMap(IO.done(_))
```

Comment:
hmm, we may have to add a `poll` to `Handoff`. Say both fibers are done and downstream is pulling, this will block ad infinitum; the contract for streams is that they should just keep on emitting `Take.end`. What do you think?

Reply:
@regiskuckaertz good point, will make it `poll` when done is true.

Reply:
Done